### PR TITLE
Initial Temperature Plug-in - Lithosphere mask

### PR DESCRIPTION
**Pull request by SophieCoulson**
New initial temperature plug in that sets a specified lithosphere temperature above the lithsphere-asthenosphere boundary, which is specified by an ascii file or a maximum lithopshere depth value. Below this the initial temperature is set as NaN. I also created an associated test.
This plug-in is to be combined with another using the 'replace if valid' operator. All the changes from #2968 are also included here. @tjhei suggested one option for fixing the test on the #2968 PR could be to modify the test to use a different plugin. Perhaps the test tests/initial_condition_lithosphere_mask tests both of these functions - I could just remove tests/initial_temperature_list_using_replace_if_valid?

### CODE REVIEW

File: include/aspect/initial_temperature/lithosphere_mask.h [hunk 1]
```diff
@@ -0,0 +1,128 @@
+/*
+  Copyright (C) 2011 - 2018 by the authors of the ASPECT code.
```
Comment (gassmoeller) [inline include/aspect/initial_temperature/lithosphere_mask.h]: switch this to `Copyright (C) 2019`

File: include/aspect/initial_temperature/lithosphere_mask.h [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+        /**
+         * Directory in which the LAB depth file is present.
+         */
+        std::string data_directory;
+
+        /**
+         * File name of the LAB depth file.
+         */
+        std::string LAB_file_name;
```
Comment (gassmoeller) [inline include/aspect/initial_temperature/lithosphere_mask.h]: All of these member variables can be moved to the `private` section further down.

File: include/aspect/initial_temperature/lithosphere_mask.h [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+        /**
+         * Return LAB depth as a function of position (latitude and longitude). For the
+         * current class, this function returns value from the text files.
+         */
+        double
+        ascii_lab (const Point<2> &position) const;
```
Comment (gassmoeller) [inline include/aspect/initial_temperature/lithosphere_mask.h]: does this function need to `public`? I.e. do you need to call it from outside the .cc file?

File: include/aspect/initial_temperature/lithosphere_mask.h [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+        /**
+         * This parameter gives the maximum depth of the lithosphere. The
+         * model be nans below this depth.
```
Comment (gassmoeller) [inline include/aspect/initial_temperature/lithosphere_mask.h]:
```suggestion
         * model returns nans below this depth.
```

File: source/initial_temperature/lithosphere_mask.cc [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+namespace aspect
+{
+  namespace InitialTemperature
+  {
+    ASPECT_REGISTER_INITIAL_TEMPERATURE_MODEL(LithosphereMask,
+                                              "lithosphere mask",
+                                              "Implementation of a model in which the initial "
+                                              "temperature is set at a specified lithosphere temperature above the "
+                                              "above the lithsphere-asthenosphere boundary (specified by an ascii file "
+                                              "or maximum lithopshere depth value). Below this the initial temperature is set as "
```
Comment (gassmoeller) [inline source/initial_temperature/lithosphere_mask.cc]:
```suggestion
                                              "or maximum lithosphere depth value). Below this the initial temperature is set as "
```

File: source/initial_temperature/lithosphere_mask.cc [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+        prm.enter_subsection ("Lithosphere Mask");
+        {
+          prm.declare_entry ("Lithosphere temperature", "1600",
+                             Patterns::Double (0),
+                             "The initial temperature within lithosphere, applied above"
+                             "the maximum lithosphere depth.");
+          prm.declare_entry ("Depth specification method", "Value",
+                             Patterns::Selection("File|Value"),
+                             "Method that is used to specify the depth of the lithosphere-asthenosphere boundary.");
+          prm.declare_entry ("Maximum lithosphere depth", "200.0",
```
Comment (gassmoeller) [inline source/initial_temperature/lithosphere_mask.cc]: This assumes to be in `km`? Please change that to `m` and document that it is interpreted as meters in the documentation text. We try to stick to SI units as much as possible.

File: include/aspect/initial_temperature/lithosphere_mask.h (hunk 4 above)
Comment (gassmoeller) [reply include/aspect/initial_temperature/lithosphere_mask.h]: Also add to the documentation that this parameter is only of relevance if `LAB_depth_source` is set to a particular value.

File: include/aspect/initial_temperature/lithosphere_mask.h [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+
+        /**
+         * This parameter gives the initial temperature set within the lithosphere.
+         */
+        double lithosphere_temperature;
+
+        /**
+         * Return the initial temperature as a function of position. For the
+         * current class, this function uses the given lithosphere temperature
+         * above max_depth and nans below max_depth.
```
Comment (gassmoeller) [inline include/aspect/initial_temperature/lithosphere_mask.h]:
```suggestion
         * above the lithosphere asthenosphere boundary and nans below.
```

File: source/initial_temperature/lithosphere_mask.cc [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+    {
+      const std::string filename = data_directory+LAB_file_name;
+
+      std::cout << "   Loading Ascii data lookup file " << filename << "." << std::endl;
+
+      lab_depths.load_file(filename,this->get_mpi_communicator());
+    }
+    }
+
+    //Read in ascii data - two dimensions so the third column is treated as data
```
Comment (gassmoeller) [inline source/initial_temperature/lithosphere_mask.cc]: This comment should go into the header file.

File: source/initial_temperature/lithosphere_mask.cc [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+    template <int dim>
+    double
+    LithosphereMask<dim>::ascii_lab (const Point<2> &position) const
+    {
+      const double lab = lab_depths.get_data(position,0)*1000; //In km
```
Comment (gassmoeller) [inline source/initial_temperature/lithosphere_mask.cc]: Either document in the class description at the bottom of this file that the file has to be in `km`, or (my preference) only allow files that input `m`, or (if you think it is important to support different units) add an input parameter that is a scale factor and is used here to multiply the value.

File: source/initial_temperature/lithosphere_mask.cc [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+      return lab;
+    }
+
+    template <int dim>
+    double
+    LithosphereMask<dim>::initial_temperature (const Point<dim> &position) const
+    {
+      double temperature;
+      double depth = this->SimulatorAccess<dim>::get_geometry_model().depth(position);
```
Comment (gassmoeller) [inline source/initial_temperature/lithosphere_mask.cc]: make this `const double ...`

File: source/initial_temperature/lithosphere_mask.cc [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+
+          if (depth <= lab_depth)
+            temperature = lithosphere_temperature;
+          else
+            temperature = std::numeric_limits<double>::quiet_NaN();
+        }
+
+      else if (LAB_depth_source == Value)
+        {
+          if (depth <= max_depth*1000)//in km
```
Comment (gassmoeller) [inline source/initial_temperature/lithosphere_mask.cc]: Same comment as above. It is easiest if everything would be in `m` then you do not need any conversion in all the files. However, if you want to keep the files in `km` this needs to be documented in the class description, and the parameter for the lithosphere thickness file.

File: source/initial_temperature/lithosphere_mask.cc [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+        }
+        prm.leave_subsection();
+      }
+      prm.leave_subsection();
+    }
+
+    template <int dim>
+    void
+
+
```
Comment (gassmoeller) [inline source/initial_temperature/lithosphere_mask.cc]: remove these empty lines

File: source/initial_temperature/lithosphere_mask.cc [hunk 13]
```diff
@@ -0,0 +1,195 @@
...
+      {
+        prm.enter_subsection ("Lithosphere Mask");
+        {
+          lithosphere_temperature   = prm.get_double ("Lithosphere temperature");
+
+          if ( prm.get("Depth specification method") == "File" )
+            {
+              LAB_depth_source = File;
+              data_directory = Utilities::expand_ASPECT_SOURCE_DIR (prm.get("Data directory"));
+              LAB_file_name   = prm.get("LAB depth filename");
```
Comment (gassmoeller) [inline source/initial_temperature/lithosphere_mask.cc]: please either align the `=` signs, or put them with a single space after the name of the variable (your choice). Just nitpicking, but the current spacing looks weird.

File: source/initial_temperature/lithosphere_mask.cc [hunk 14]
```diff
@@ -0,0 +1,195 @@
...
+            }
+          else if ( prm.get("Depth specification method") == "Value" )
+            {
+              LAB_depth_source = Value;
+              max_depth           = prm.get_double ("Maximum lithosphere depth");
```
Comment (gassmoeller) [inline source/initial_temperature/lithosphere_mask.cc]: same here

File: source/initial_temperature/lithosphere_mask.cc [hunk 15]
```diff
@@ -0,0 +1,195 @@
...
+            }
+
+
+
```
Comment (gassmoeller) [inline source/initial_temperature/lithosphere_mask.cc]: please remove the unnecessary empty lines

File: source/initial_temperature/lithosphere_mask.cc [hunk 16]
```diff
@@ -0,0 +1,195 @@
...
+// explicit instantiations
+namespace aspect
+{
+  namespace InitialTemperature
+  {
+    ASPECT_REGISTER_INITIAL_TEMPERATURE_MODEL(LithosphereMask,
+                                              "lithosphere mask",
+                                              "Implementation of a model in which the initial "
+                                              "temperature is set at a specified lithosphere temperature above the "
```
Comment (gassmoeller) [inline source/initial_temperature/lithosphere_mask.cc]:
```suggestion
                                              "temperature is set to a specified lithosphere temperature above the "
```

File: source/initial_temperature/lithosphere_mask.cc [hunk 17]
```diff
@@ -0,0 +1,195 @@
...
+                                              "above the lithsphere-asthenosphere boundary (specified by an ascii file "
```
Comment (gassmoeller) [inline source/initial_temperature/lithosphere_mask.cc]: double `above the`, and `lithsphere` -> `lithosphere`

File: source/initial_temperature/patch_on_S40RTS.cc [hunk 18]
```diff
@@ -39,7 +39,7 @@ namespace aspect
     void
     PatchOnS40RTS<dim>::initialize ()
     {
-      Utilities::AsciiDataInitial<dim>::initialize(1);
+      this->Utilities::AsciiDataInitial<dim>::initialize(1);
```
Comment (gassmoeller) [inline source/initial_temperature/patch_on_S40RTS.cc]: Is this necessary, or an accidental change?

Comment (SophieCoulson) [reply source/initial_temperature/patch_on_S40RTS.cc]: Wolfgang preferred this notation but its not strictly necessary - sorry, it maybe should have been a separate PR.

**Comment by SophieCoulson** [comment]
Thanks Rene, I've made all the changes requested. I've also deleted the test for the 'replace if valid' operator, this is now tested in the tests/initial_condition_lithosphere_mask test. This means we never have to say 'nan' in the parameter file, so the tests will now work on microsoft. @bangerth @tjhei Are you happy with this solution if I delete the other PR?

File: include/aspect/initial_temperature/lithosphere_mask.h [hunk 19]
```diff
@@ -56,8 +56,32 @@ namespace aspect
...
+         * Read the parameters this class declares from the parameter file.
+         */
+        void
+        parse_parameters (ParameterHandler &prm);
+
+      private:
+        Utilities::AsciiDataLookup<2> lab_depths;
+
+        /**
+                 * Directory in which the LAB depth file is present.
```
Comment (gassmoeller) [inline include/aspect/initial_temperature/lithosphere_mask.h]: You will need to fix the formatting of the comments.

File: include/aspect/initial_temperature/lithosphere_mask.h [hunk 20]
```diff
@@ -56,8 +56,32 @@ namespace aspect
...
+        declare_parameters (ParameterHandler &prm);
+
+        /**
+         * Read the parameters this class declares from the parameter file.
+         */
+        void
+        parse_parameters (ParameterHandler &prm);
+
+      private:
+        Utilities::AsciiDataLookup<2> lab_depths;
```
Comment (gassmoeller) [inline include/aspect/initial_temperature/lithosphere_mask.h]: Add some documentation for this variable.

File: include/aspect/initial_temperature/lithosphere_mask.h [hunk 21]
```diff
@@ -67,14 +91,17 @@ namespace aspect
 
         /**
          * Return LAB depth as a function of position (latitude and longitude). For the
-         * current class, this function returns value from the text files.
+         * current class, this function returns value from the text file. We read in in
```
Comment (gassmoeller) [inline include/aspect/initial_temperature/lithosphere_mask.h]: `in in` -> `in`

File: source/initial_temperature/lithosphere_mask.cc [hunk 22]
```diff
@@ -115,7 +114,7 @@ namespace aspect
           prm.declare_entry ("Depth specification method", "Value",
                              Patterns::Selection("File|Value"),
                              "Method that is used to specify the depth of the lithosphere-asthenosphere boundary.");
-          prm.declare_entry ("Maximum lithosphere depth", "200.0",
+          prm.declare_entry ("Maximum lithosphere depth", "200000.0",
                              Patterns::Double (0),
                              "The maximum depth of the lithosphere. The model will be "
                              "NaNs below this depth.");
```
Comment (gassmoeller) [inline source/initial_temperature/lithosphere_mask.cc]: Add a quick note in the description: `Units: m.`

**Comment by SophieCoulson** [comment]
Done!

**Comment by gassmoeller** [comment]
You will need to update the test output from the tester. Otherwise ready.

**Comment by SophieCoulson** [comment]
I've updated the test output - hopefully its good to go.

**Comment by gassmoeller** [comment]
Looks good, thanks!

**Pull request merged**